tests/test_rtdc_config.py

```diff
@@ -81,6 +85,284 @@ def test_config_update():
+    ds.config["user"]["channel D"] = "a string"
+    ds.config["user"]["channel E"] = {"a key": 23}
+    ds.config["user"]["channel F"] = [8, 55.4, True, "a word"]
+    ds.config["user"]["channel G"] = (5, 12.3, False, "a word")
+
+    assert len(ds.config["user"]) == 7
+    assert isinstance(ds.config["user"]["channel A"], bool)
+    assert isinstance(ds.config["user"]["channel B"], float)
+    assert isinstance(ds.config["user"]["channel C"], int)
+    assert isinstance(ds.config["user"]["channel D"], str)
+    assert isinstance(ds.config["user"]["channel E"], dict)
+    assert isinstance(ds.config["user"]["channel F"], list)
+    assert isinstance(ds.config["user"]["channel G"], tuple)
+
+
+def test_user_section_exists():
+    """Check that the user config section exists"""
+    ds = new_dataset(retrieve_data("rtdc_data_hdf5_rtfdc.zip"))
+    assert ds.config["user"] == {}
+    # show that nonsense sections don't exist
+    with pytest.raises(KeyError):
+        ds.config["Oh I seem to have lost my key"]
+
+
+def test_user_section_set_and_overwrite():
+    """Add information to the user section of config via dict.__setitem__"""
+    ds = new_dataset(retrieve_data("rtdc_data_hdf5_rtfdc.zip"))
+    ds.config["user"]["some metadata"] = 42
+    assert ds.config["user"] == {"some metadata": 42}
+    metadata = {"more metadata": True}
+    ds.config.update({"user": metadata})
+    assert ds.config["user"] == {"some metadata": 42,
+                                 "more metadata": True}
+    # overwrite the previous keys and valus
+    ds.config["user"] = {}
+    assert ds.config["user"] == {}
+
+
+def test_user_section_set_save_reload_empty_dict():
+    """The 'user' config section as an empty dict will not save"""
+    h5path = retrieve_data("rtdc_data_hdf5_rtfdc.zip")
+    with new_dataset(h5path) as ds:
+        ds.config.update({"user": {}})
+        expath = h5path.with_name("exported.rtdc")
+        ds.export.hdf5(expath, features=ds.features_innate)
+    # fails for hdf5
+    with pytest.raises(KeyError):
+        with h5py.File(expath, "r") as h5:
+            assert h5.attrs["user:"] == {}
+    # works for dclab because "user" added when checked
+    with new_dataset(expath) as ds2:
+        assert ds2.config["user"] == {}
+
+
+def test_user_section_set_save_reload_empty_key():
+    """Empty 'user' section key value allowed"""
+    h5path = retrieve_data("rtdc_data_hdf5_rtfdc.zip")
+    with new_dataset(h5path) as ds:
+        ds.config.update({"user": {"": " "}})
+        expath = h5path.with_name("exported.rtdc")
+        ds.export.hdf5(expath, features=ds.features_innate)
+    # make sure that worked
+    with h5py.File(expath, "r") as h5:
+        assert h5.attrs["user:"] == " "
+    # now check again with dclab
+    with new_dataset(expath) as ds2:
+        assert ds2.config["user"] == {"": " "}
+
+
+@pytest.mark.parametrize("user_config", [{"": ""}, {" ": ""}])
+def test_user_section_set_save_reload_fails(user_config):
```

Comment:
Welcome to the word of empty strings!
Well `config.py` does not allow the value to be an empty string. That shows why `{"": ""}` and `{" ": ""}` are not allowed. So these cases are already covered. 

We could indeed add a check for `{"": " "}`. It shouldn't be allowed...

But what about `{" ": " "}`? Will we go to the bother of checking that there are more than one character that isn't a single space?

Reply:
I think it would be safe to at least to a `.strip()` for every key (not for the value) and make sure the result is not empty. This check is a convenience check for users to prevent bad things from happening.